File: aoc22/java/src/test/java/tech/dobler/aoc22/Day11Test.java
package tech.dobler.aoc22;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class Day11Test {

    final String testInput = """
            Monkey 0:
              Starting items: 79, 98
              Operation: new = old * 19
              Test: divisible by 23
                If true: throw to monkey 2
                If false: throw to monkey 3
                        
            Monkey 1:
              Starting items: 54, 65, 75, 74
              Operation: new = old + 6
              Test: divisible by 19
                If true: throw to monkey 2
                If false: throw to monkey 0
                        
            Monkey 2:
              Starting items: 79, 60, 97
              Operation: new = old * old
              Test: divisible by 13
                If true: throw to monkey 1
                If false: throw to monkey 3
                        
            Monkey 3:
              Starting items: 74
              Operation: new = old + 3
              Test: divisible by 17
                If true: throw to monkey 0
                If false: throw to monkey 1
            """;
    private Day11 day;

    @BeforeEach
    void setUp() {
        day = new Day11();
    }

    private static List<Day11.Monkey> readInput() {
        return Util.splitByDoubleNewLine(Util.readInput(11), Day11.Monkey::from).toList();
    }

    private List<Day11.Monkey> parseTestInput() {
        return Util.splitByDoubleNewLine(testInput, Day11.Monkey::from).toList();
    }

    @Test
    void testTransformation() {
        final var parsedTestInput = parseTestInput();
        Util.printfln("Test input contains: %d entries", parsedTestInput.size());
        assertThat(parsedTestInput).isNotEmpty().hasSize(4);

        final var input = readInput();
        Util.printfln("Input contains: %d entries", input.size());
        assertThat(input).isNotEmpty().hasSize(8);
    }

    @Test
    void part1() {
        assertThat(day.part1(parseTestInput())).isEqualTo(10_605L);

        assertThat(day.part1(readInput())).isEqualTo(90_882L);
    }
    @Test
    void part2() {
        assertThat(day.part2(parseTestInput())).isEqualTo(2_713_310_158L);

        assertThat(day.part2(readInput())).isEqualTo(30_893_109_657L);
    }
}

File: aoc21/Kotlin/src/test/kotlin/day11/Day11Test.kt
package day11

import org.junit.jupiter.api.Test
import util.readInput
import kotlin.test.assertEquals

class Day11Test {

    private val path = "11_test"

    @Test
    fun `parse input`() {
        val input = readInput(path, ::mapInput);
        assertEquals(
            listOf(
                listOf(5, 4, 8, 3, 1, 4, 3, 2, 2, 3),
                listOf(2, 7, 4, 5, 8, 5, 4, 7, 1, 1),
                listOf(5, 2, 6, 4, 5, 5, 6, 1, 7, 3),
                listOf(6, 1, 4, 1, 3, 3, 6, 1, 4, 6),
                listOf(6, 3, 5, 7, 3, 8, 5, 4, 7, 8),
                listOf(4, 1, 6, 7, 5, 2, 4, 6, 4, 5),
                listOf(2, 1, 7, 6, 8, 4, 1, 7, 2, 1),
                listOf(6, 8, 8, 2, 8, 8, 1, 1, 3, 4),
                listOf(4, 8, 4, 6, 8, 4, 8, 5, 5, 4),
                listOf(5, 2, 8, 3, 7, 5, 1, 5, 2, 6)
            ), input
        )
    }

    @Test
    fun `part 1 example`() {
        val input = readInput(path, ::mapInput);
        assertEquals(1_656, part1(input))
    }

    @Test
    fun `part 2 example`() {
        val input = readInput(path, ::mapInput);
        assertEquals(Int.MIN_VALUE, part2(input))
    }
}